crates/authentication/src/sys/linux/mod.rs

```diff
@@ -0,0 +1,169 @@
+use zbus_polkit::policykit1::{AuthorityProxyBlocking, AuthorizationResult, CheckAuthorizationFlags, Subject};
+
+use crate::{Error, Result, Text};
+
+pub(crate) type RawContext = ();
+
+#[derive(Debug)]
+pub(crate) struct Context;
+
+impl Context {
+    #[inline]
+    pub(crate) fn new(_: RawContext) -> Self {
+        Self
+    }
+
+    pub(crate) fn authenticate<F>(
+        &self,
+        _message: Text, // _message is unused on Linux, see reasons below:  auth.check_authorization();
+        policy: &Policy,
+        callback: F,
+    ) -> Result<()>
+    where
+        F: Fn(Result<()>) + Send + 'static,
+    {
+        let action_id = policy.action_id;
+        // Here, we perform a simple and direct thread creation because the current calls are infrequent.
+        thread::Builder::new()
+            .name(String::from("robius-authentication-polkit"))
+            .spawn(move || {
+                let res = do_polkit_check(action_id);
+                callback(res);
+            })
+            .map_err(|_| Error::Unavailable)?;
+        Ok(())
+    }
+}
+
+static SYSTEM_CONNECTION: OnceLock<Result<Connection>> = OnceLock::new(); // Cached system bus connection
+static AUTHORITY_PROXY: OnceLock<Result<AuthorityProxyBlocking<'static>>> = OnceLock::new(); // Cached authority proxy
+
+pub(crate) fn get_system_connection() -> Result<Connection> {
+    let r: &Result<Connection> = SYSTEM_CONNECTION.get_or_init(|| {
+        Connection::system()
+            .map_err(|_| Error::Unavailable)
+    });
+
+    match r {
+        Ok(conn) => Ok(conn.clone()),
+        Err(e) => Err(e.clone()),
+    }
+}
+
+pub(crate) fn get_authority_proxy() -> Result<AuthorityProxyBlocking<'static>> {
+    let r: &Result<AuthorityProxyBlocking<'static>> = AUTHORITY_PROXY.get_or_init(|| {
+        let conn = get_system_connection()?;
+        AuthorityProxyBlocking::new(&conn).map_err(|e| {
+            let _ = e;
+            Error::Unavailable
+        })
+    });
+
+    match r {
+        Ok(p) => Ok(p.clone()),
+        Err(e) => Err(e.clone()),
+    }
+}
+
+fn do_polkit_check(action_id: &'static str) -> Result<()> {
+    // Get authority proxy, and it's cached for future use.
+    let auth = get_authority_proxy()?;
+
+    let pid = std::process::id() as u32;
+    let subject = Subject::new_for_owner(pid, None, None)
+        .map_err(|_| Error::Unavailable)?;
+
+    // If details is non-empty then the request will fail with POLKIT_ERROR_FAILED unless the process doing the check itsef is sufficiently authorized (e.g. running as uid 0).
+    let result = auth
+        .check_authorization(
+            &subject,
+            action_id,
+            &HashMap::new(), 
+            CheckAuthorizationFlags::AllowUserInteraction.into(),
+            "",
+        )
+        .map_err(|e| map_polkit_error(e.to_string()))?;
+
+    map_authorization_result(result)
+}
+
+fn map_authorization_result(result: AuthorizationResult) -> Result<()> {
```

Comment:
no need to make this a separate function, just combine it into the above function.